File: src/free_claude_code/runtime/codex_browser_helper.py
"""Approved-helper adapter for the installed Codex browser plugin."""

import json
import os
import shutil
import subprocess
import threading
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from free_claude_code.application.capabilities import Capability
from free_claude_code.application.helpers import ApprovedHelper

CODEX_BROWSER_HELPER_ID = "codex-browser"
CODEX_BROWSER_PROVIDER_FAMILY = "browser"
DEFAULT_BROWSER_OUTPUT_BYTES = 24 * 1024 * 1024
BROWSER_FAMILIES = frozenset({"chrome", "edge"})
BROWSER_OPERATIONS = frozenset(
    {
        "list_tabs",
        "claim_tab",
        "new_tab",
        "selected_tab",
        "tab_info",
        "goto",
        "snapshot",
        "click",
        "set_value",
        "type_text",
        "press_key",
        "scroll",
        "screenshot",
        "reload",
        "back",
        "forward",
        "mark_handoff",
        "mark_deliverable",
        "close_tab",
    }
)
READ_ONLY_BROWSER_OPERATIONS = frozenset(
    {"list_tabs", "selected_tab", "tab_info", "snapshot", "screenshot"}
)
MUTATING_BROWSER_OPERATIONS = BROWSER_OPERATIONS - READ_ONLY_BROWSER_OPERATIONS
_ENV_ALLOWLIST = frozenset(
    {
        "CODEX_HOME",
        "HOME",
        "LANG",
        "LC_ALL",
        "LOGNAME",
        "NODE_PATH",
        "PATH",
        "SHELL",
        "TEMP",
        "TMP",
        "TMPDIR",
        "USER",
        "XDG_CACHE_HOME",
        "XDG_CONFIG_HOME",
        "XDG_RUNTIME_DIR",
    }
)


class CodexBrowserHelperError(RuntimeError):
    """The installed Codex browser bridge could not serve one bounded call."""


class CodexBrowserHelperAdapter:
    """Own one warm Node bridge to Codex's installed browser plugin."""

    def __init__(
        self,
        *,
        family: str = "chrome",
        plugin_root: Path | None = None,
        session_id: str = "fcc-browser",
    ) -> None:
        if family not in BROWSER_FAMILIES:
            raise ValueError("browser family must be chrome or edge")
        if plugin_root is not None and not plugin_root.is_absolute():
            raise ValueError("Codex browser plugin root must be absolute")
        if not session_id.strip():
            raise ValueError("session_id is required")
        self._family = family
        self._plugin_root = plugin_root
        self._session_id = session_id
        self._process: subprocess.Popen[str] | None = None
        self._process_lock = threading.Lock()
        self._request_lock = threading.Lock()
        self._next_id = 1

    def close(self) -> None:
        """Terminate only the bridge process owned by this adapter."""

        with self._process_lock:
            process = self._process
            self._process = None
        if process is None:
            return
        if process.poll() is None:
            process.terminate()
            try:
                process.wait(timeout=0.5)
            except subprocess.TimeoutExpired:
                process.kill()
                process.wait(timeout=0.5)
        if process.stdin is not None:
            process.stdin.close()
        if process.stdout is not None:
            process.stdout.close()

    def execute(
        self,
        operation: str,
        arguments: Mapping[str, Any],
        cancel_event: threading.Event,
    ) -> Mapping[str, Any]:
        """Execute one allowlisted browser operation through the installed plugin."""

        if operation not in BROWSER_OPERATIONS:
            raise CodexBrowserHelperError(
                f"unsupported Codex browser helper operation: {operation}"
            )
        if cancel_event.is_set():
            raise CodexBrowserHelperError(
                "Codex browser helper cancelled before dispatch"
            )

        with self._request_lock:
            process = self._get_process()
            finished = threading.Event()

            def cancel_watcher() -> None:
                while not finished.wait(timeout=0.05):
                    if cancel_event.is_set():
                        self.close()
                        return

            watcher = threading.Thread(
                target=cancel_watcher,
                name="fcc-codex-browser-cancel",
                daemon=True,
            )
            watcher.start()
            try:
                return self._call(process, operation, arguments)
            finally:
                finished.set()
                watcher.join(timeout=0.2)

    def approved_helper(
        self,
        *,
        max_output_bytes: int = DEFAULT_BROWSER_OUTPUT_BYTES,
    ) -> ApprovedHelper:
        """Return deterministic #30/#104 metadata for this local helper."""

        return ApprovedHelper(
            helper_id=CODEX_BROWSER_HELPER_ID,
            provider_family=CODEX_BROWSER_PROVIDER_FAMILY,
            capabilities=frozenset(
                {Capability.SEMANTIC_BROWSER_CONTROL, Capability.SCREENSHOT_VISION}
            ),
            execute=self.execute,
            local=True,
            billable=False,
            max_output_bytes=max_output_bytes,
            mutating_operations=MUTATING_BROWSER_OPERATIONS,
        )

    def _get_process(self) -> subprocess.Popen[str]:
        with self._process_lock:
            process = self._process
            if process is not None and process.poll() is None:
                return process

            node = shutil.which("node")
            if node is None:
                raise CodexBrowserHelperError(
                    "Node.js is required for Codex browser use"
                )
            script = Path(__file__).with_name("codex_browser_helper.mjs")
            if not script.is_file():
                raise CodexBrowserHelperError(
                    "packaged Codex browser helper is missing"
                )

            process = subprocess.Popen(
                [node, str(script)],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
                text=True,
                encoding="utf-8",
                bufsize=1,
                env=self._child_environment(),
            )
            if process.stdin is None or process.stdout is None:
                process.kill()
                raise CodexBrowserHelperError(
                    "failed to open Codex browser helper pipes"
                )
            self._process = process
            return process

    def _call(
        self,
        process: subprocess.Popen[str],
        operation: str,
        arguments: Mapping[str, Any],
    ) -> dict[str, Any]:
        if process.stdin is None or process.stdout is None:
            raise CodexBrowserHelperError("Codex browser helper pipes are unavailable")
        request_id = self._next_id
        self._next_id += 1
        request = {
            "id": request_id,
            "operation": operation,
            "arguments": dict(arguments),
        }
        try:
            encoded = json.dumps(request, separators=(",", ":"), ensure_ascii=False)
        except (TypeError, ValueError) as error:
            raise CodexBrowserHelperError(
                "browser arguments must be JSON-compatible"
            ) from error

        try:
            process.stdin.write(encoded + "\n")
            process.stdin.flush()
            line = process.stdout.readline()
        except (BrokenPipeError, OSError) as error:
            self.close()
            raise CodexBrowserHelperError(
                "Codex browser helper transport failed"
            ) from error
        if not line:
            self.close()
            raise CodexBrowserHelperError(
                "Codex browser helper exited without a response"
            )
        try:
            response = json.loads(line)
        except json.JSONDecodeError as error:
            self.close()
            raise CodexBrowserHelperError(
                "Codex browser helper returned invalid JSON"
            ) from error
        if not isinstance(response, Mapping) or response.get("id") != request_id:
            self.close()
            raise CodexBrowserHelperError(
                "Codex browser helper response correlation failed"
            )
        if response.get("ok") is not True:
            raise CodexBrowserHelperError("Codex browser operation failed")
        result = response.get("result")
        if not isinstance(result, Mapping):
            raise CodexBrowserHelperError(
                "Codex browser helper returned non-object output"
            )
        return dict(result)

    def _child_environment(self) -> dict[str, str]:
        environment = {
            key: value for key, value in os.environ.items() if key in _ENV_ALLOWLIST
        }
        environment["FCC_CODEX_BROWSER_FAMILY"] = self._family
        environment["FCC_CODEX_BROWSER_SESSION_ID"] = self._session_id
        if self._plugin_root is not None:
            environment["FCC_CODEX_BROWSER_PLUGIN_ROOT"] = str(self._plugin_root)
        return environment


__all__ = [
    "BROWSER_FAMILIES",
    "BROWSER_OPERATIONS",
    "CODEX_BROWSER_HELPER_ID",
    "CODEX_BROWSER_PROVIDER_FAMILY",
    "MUTATING_BROWSER_OPERATIONS",
    "READ_ONLY_BROWSER_OPERATIONS",
    "CodexBrowserHelperAdapter",
    "CodexBrowserHelperError",
]

File: src/free_claude_code/runtime/codex_browser_helper.mjs
// Bounded bridge to the installed Codex Chrome/Edge browser plugin.
// Adapted from deepcoldy/botmux@eff1953a66fe6054f47d9311c719ab73c4ed6f1d (MIT).

import { existsSync, readdirSync, realpathSync } from "node:fs";
import { createConnection } from "node:net";
import { homedir, platform, tmpdir } from "node:os";
import { basename, isAbsolute, join, resolve } from "node:path";
import { createInterface } from "node:readline";
import { pathToFileURL } from "node:url";

const MAX_TEXT_BYTES = 512 * 1024;
const MAX_SCREENSHOT_BYTES = 16 * 1024 * 1024;
const MAX_URL_BYTES = 16 * 1024;
const MAX_VALUE_BYTES = 100 * 1024;
const FAMILY = process.env.FCC_CODEX_BROWSER_FAMILY || "chrome";
const EXPLICIT_PLUGIN_ROOT = process.env.FCC_CODEX_BROWSER_PLUGIN_ROOT || undefined;
const SESSION_ID = process.env.FCC_CODEX_BROWSER_SESSION_ID || "fcc-browser";

if (FAMILY !== "chrome" && FAMILY !== "edge") {
  throw new Error("FCC_CODEX_BROWSER_FAMILY must be chrome or edge");
}

function isCompletePluginRoot(root) {
  return existsSync(join(root, "scripts/browser-client.mjs"))
    && existsSync(join(root, "scripts/browser-service.mjs"));
}

function compareVersionNames(left, right) {
  const leftParts = basename(left).split(".").map(Number);
  const rightParts = basename(right).split(".").map(Number);
  for (let index = 0; index < Math.max(leftParts.length, rightParts.length); index += 1) {
    const delta = (leftParts[index] || 0) - (rightParts[index] || 0);
    if (delta) return delta;
  }
  return left.localeCompare(right);
}

function resolvePluginRoot() {
  if (EXPLICIT_PLUGIN_ROOT && !isAbsolute(EXPLICIT_PLUGIN_ROOT)) {
    throw new Error("Codex browser plugin root must be absolute");
  }
  const codexHome = process.env.CODEX_HOME?.trim() || join(homedir(), ".codex");
  const cache = join(codexHome, "plugins", "cache", "openai-bundled", "chrome");
  const explicit = EXPLICIT_PLUGIN_ROOT ? resolve(EXPLICIT_PLUGIN_ROOT) : undefined;
  if (explicit) {
    const root = realpathSync(explicit);
    if (!isCompletePluginRoot(root)) throw new Error("Codex browser plugin is incomplete");
    return root;
  }
  const latest = join(cache, "latest");
  if (isCompletePluginRoot(latest)) return realpathSync(latest);
  let entries;
  try {
    entries = readdirSync(cache, { withFileTypes: true });
  } catch {
    throw new Error(`Codex browser plugin is not installed under ${cache}`);
  }
  const candidates = entries
    .filter((entry) => entry.isDirectory() || entry.isSymbolicLink())
    .map((entry) => join(cache, entry.name))
    .filter(isCompletePluginRoot)
    .sort((left, right) => compareVersionNames(right, left));
  if (!candidates[0]) throw new Error(`Codex browser plugin is not installed under ${cache}`);
  return realpathSync(candidates[0]);
}

function connectNativePipe(path) {
  return new Promise((resolveConnection, reject) => {
    const socket = createConnection(path);
    const onError = (error) => reject(error);
    socket.once("error", onError);
    socket.once("connect", () => {
      socket.off("error", onError);
      resolveConnection(socket);
    });
  });
}

function requireString(value, field, maxBytes) {
  if (typeof value !== "string" || Buffer.byteLength(value, "utf8") > maxBytes) {
    throw new Error(`${field} must be a string no larger than ${maxBytes} bytes`);
  }
  return value;
}

function requireNonEmptyString(value, field, maxBytes) {
  const text = requireString(value, field, maxBytes);
  if (!text) throw new Error(`${field} must not be empty`);
  return text;
}

function requireIndex(value, field) {
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${field} must be a non-negative integer`);
  }
  return value;
}

function requireFinite(value, field) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`${field} must be finite`);
  }
  return value;
}

function boundedText(value) {
  let text = typeof value === "string" ? value : JSON.stringify(value);
  if (Buffer.byteLength(text, "utf8") > MAX_TEXT_BYTES) {
    text = `${Buffer.from(text, "utf8").subarray(0, MAX_TEXT_BYTES).toString("utf8")}\n[truncated by FCC]`;
  }
  return { text };
}

let browser;
let initialization;
const claimedTabs = new Map();
const configStore = new Map();
let runtimeShim;

function withRequestMetadata(request) {
  if (request.method !== "execute" || runtimeShim == null) return request;
  const params = request.params;
  if (params == null || typeof params !== "object" || Array.isArray(params)) {
    return request;
  }
  return {
    ...request,
    params: { ...params, requestMeta: runtimeShim.requestMeta },
  };
}

async function initialize(initialRequest) {
  const root = resolvePluginRoot();
  const service = await import(pathToFileURL(join(root, "scripts/browser-service.mjs")).href);
  const client = await import(pathToFileURL(join(root, "scripts/browser-client.mjs")).href);
  runtimeShim = {
    env: { ...process.env },
    cwd: process.cwd(),
    homeDir: homedir(),
    tmpDir: tmpdir(),
    platform: platform(),
    requestMeta: {},
    nativePipe: { createConnection: connectNativePipe },
    config: {
      read: async () => ({}),
      readRequirements: async () => ({}),
      readToml: async (path) => configStore.get(path) ?? {},
      writeToml: async (path, value) => configStore.set(path, value),
    },
    createElicitation: async () => ({ action: "cancel" }),
    setResponseMeta: () => {},
    addAfterSubmittedCodeHook: () => {},
    emitContentItem: () => {},
    emitImage: () => {},
    write: () => {},
    fetch: globalThis.fetch,
    rpc: undefined,
  };
  updateMetadata(initialRequest);
  globalThis.nodeRepl = runtimeShim;
  runtimeShim.rpc = async (name, request) => {
    if (name !== "browser") throw new Error(`unsupported trusted service: ${name}`);
    return service.handleRpc(withRequestMetadata(request));
  };
  const agent = await client.setupBrowserRuntime();
  browser = await agent.browsers.get(FAMILY);
  await browser.nameSession(`fcc-${SESSION_ID.slice(0, 24)}`).catch(() => {});
  return browser;
}

async function getBrowser(initialRequest) {
  if (browser) return browser;
  initialization ??= initialize(initialRequest);
  return initialization;
}

function updateMetadata(request) {
  if (!runtimeShim) return;
  runtimeShim.requestMeta = {
    "x-codex-turn-metadata": JSON.stringify({
      session_id: SESSION_ID,
      thread_id: request.thread_id || SESSION_ID,
      turn_id: request.turn_id || String(request.id ?? "fcc"),
      thread_source: "fcc",
    }),
  };
}

async function describeTab(tab) {
  const [title, url] = await Promise.all([tab.title(), tab.url()]);
  return { id: tab.id, ...(title ? { title } : {}), ...(url ? { url } : {}) };
}

async function getTab(tabId) {
  const cached = claimedTabs.get(tabId);
  if (cached) return cached;
  const activeBrowser = await getBrowser();
  const tab = await activeBrowser.tabs.get(tabId);
  claimedTabs.set(tab.id, tab);
  return tab;
}

async function execute(request) {
  const input = request.arguments;
  if (!input || typeof input !== "object" || Array.isArray(input)) {
    throw new Error("arguments must be an object");
  }
  const operation = requireNonEmptyString(request.operation, "operation", 64);
  const activeBrowser = await getBrowser(request);
  updateMetadata(request);

  if (operation === "list_tabs") {
    return { family: FAMILY, tabs: await activeBrowser.user.openTabs() };
  }
  if (operation === "claim_tab") {
    const tabId = requireNonEmptyString(input.tab_id, "tab_id", 256);
    const info = (await activeBrowser.user.openTabs()).find((tab) => tab.id === tabId);
    if (!info) throw new Error(`user tab is no longer available: ${tabId}`);
    const tab = await activeBrowser.user.claimTab(info);
    claimedTabs.set(tabId, tab);
    claimedTabs.set(tab.id, tab);
    return describeTab(tab);
  }
  if (operation === "new_tab") {
    const tab = await activeBrowser.tabs.new();
    claimedTabs.set(tab.id, tab);
    return describeTab(tab);
  }
  if (operation === "selected_tab") {
    const tab = await activeBrowser.tabs.selected();
    if (!tab) return { tab: null };
    claimedTabs.set(tab.id, tab);
    return describeTab(tab);
  }

  const tab = await getTab(requireNonEmptyString(input.tab_id, "tab_id", 256));
  switch (operation) {
    case "tab_info":
      return describeTab(tab);
    case "goto":
      await tab.goto(requireNonEmptyString(input.url, "url", MAX_URL_BYTES));
      return describeTab(tab);
    case "snapshot":
      return boundedText(await tab.ax.get("state", { disableDiffing: input.disable_diffing === true }));
    case "click":
      await tab.ax.click(requireIndex(input.element_index, "element_index"));
      return { ok: true };
    case "set_value":
      await tab.ax.setValue(
        requireIndex(input.element_index, "element_index"),
        requireString(input.value, "value", MAX_VALUE_BYTES),
      );
      return { ok: true };
    case "type_text":
      await tab.ax.typeText(requireString(input.value, "value", MAX_VALUE_BYTES));
      return { ok: true };
    case "press_key":
      await tab.ax.pressKey(requireNonEmptyString(input.key, "key", 256));
      return { ok: true };
    case "scroll": {
      if (!["up", "down", "left", "right"].includes(input.direction)) {
        throw new Error("direction must be up, down, left, or right");
      }
      const target = input.element_index !== undefined
        ? requireIndex(input.element_index, "element_index")
        : [requireFinite(input.x, "x"), requireFinite(input.y, "y")];
      const pages = input.pages === undefined ? undefined : requireFinite(input.pages, "pages");
      if (pages !== undefined && (pages <= 0 || pages > 20)) {
        throw new Error("pages must be > 0 and <= 20");
      }
      await tab.ax.scroll(target, input.direction, pages);
      return { ok: true };
    }
    case "screenshot": {
      const bytes = await tab.screenshot({ fullPage: input.full_page === true });
      if (bytes.byteLength > MAX_SCREENSHOT_BYTES) {
        throw new Error(`screenshot exceeds ${MAX_SCREENSHOT_BYTES} bytes`);
      }
      return { media_type: "image/jpeg", image_base64: Buffer.from(bytes).toString("base64") };
    }
    case "reload":
      await tab.reload(); return { ok: true };
    case "back":
      await tab.back(); return { ok: true };
    case "forward":
      await tab.forward(); return { ok: true };
    case "mark_handoff":
      await tab.markHandoff(); return { ok: true };
    case "mark_deliverable":
      await tab.markDeliverable(); return { ok: true };
    case "close_tab":
      await tab.close(); claimedTabs.delete(tab.id); return { ok: true };
    default:
      throw new Error(`unsupported browser operation: ${operation}`);
  }
}

function reply(message) {
  process.stdout.write(`${JSON.stringify(message)}\n`);
}

const input = createInterface({ input: process.stdin, crlfDelay: Infinity });
input.on("line", async (line) => {
  if (Buffer.byteLength(line, "utf8") > MAX_VALUE_BYTES * 2) {
    reply({ id: null, ok: false, error: "request exceeds input bound" });
    return;
  }
  let request;
  try {
    request = JSON.parse(line);
    const result = await execute(request);
    reply({ id: request.id ?? null, ok: true, result });
  } catch (error) {
    reply({ id: request?.id ?? null, ok: false, error: error instanceof Error ? error.message : String(error) });
  }
});
